app/gui/view/component-browser/breadcrumbs/src/lib.rs

```diff
@@ -0,0 +1,478 @@
+    /// the user always sees the right (most important) part of the breadcrumbs.
+    fn update_layout(&self, content_size: Vector2, size: Vector2) {
+        self.mask.size.set(size);
+        let grid_view_center = Vector2(size.x / 2.0, -size.y / 2.0);
+        self.mask.set_position_xy(grid_view_center);
+        // Additional padding is added to avoid rare glitches when the last entry is
+        // cropped because it is placed right on the border of the viewport. Even 1px seems
+        // enough, but we add a bit more to be sure.
+        let padding = 10.0;
+        let offset = self.offset(content_size, size);
+        let right = offset + size.x + padding;
+        let vp = Viewport { top: 0.0, bottom: -size.y, left: offset, right };
+        self.grid.set_viewport(vp);
+    }
+
+    /// Calculate an offset of the breadcrumbs' content. If the content does not fit into [`size`],
+    /// we shift it left so that the user always sees the right (most important) part of the
+    /// breadcrumbs. Also, we keep the selected breadcrumb visible and easily reachable by
+    /// placing it near the center of the viewport.
+    fn offset(&self, content_size: Vector2, size: Vector2) -> f32 {
+        let selected_col = self.grid.entry_selected.value().map(|(_, col)| col).unwrap_or(0);
+        let last_col = self.column_of_the_last_entry().unwrap_or(0);
+        let entry_pos = self.grid.entry_position(1, selected_col).x;
+        let selected_is_not_last = selected_col < last_col;
+        let scrolling_threshold_pos = content_size.x - size.x * SCROLLING_THRESHOLD;
+        let should_scroll = entry_pos < scrolling_threshold_pos;
+        let content_truncated = content_size.x > size.x;
+        let content_right = if content_truncated && selected_is_not_last && should_scroll {
+            entry_pos + SCROLLING_THRESHOLD * size.x
+        } else {
+            content_size.x
+        };
+        let viewport_right = content_right.min(size.x);
+        content_right - viewport_right
+    }
+
+    /// A model for the specific entry. Every second entry of the grid view is an actual
+    /// breadcrumb. They are separated by the [`entry::Model::Separator`] entries and can have an
+    /// optional [`entry::Model::Ellipsis`] icon as the last entry (if [`show_ellipsis`] is true).
+    fn entry_model(
+        entries: &Entries,
+        col: Col,
+        show_ellipsis: bool,
+        number_of_cols: Col,
+    ) -> entry::Model {
+        let is_last = col == number_of_cols - 1;
+        let is_not_first = col != 0;
+        let is_separator_index = col % 2 == 1;
+        if show_ellipsis && is_last && is_not_first {
+            entry::Model::Ellipsis
+        } else if is_separator_index {
+            entry::Model::Separator
+        } else if let Some(entry) = entries.borrow().get(col / 2) {
+            entry::Model::Text(entry.0.clone_ref())
+        } else {
+            tracing::error!("Requested entry is missing in the breadcrumbs ({col})");
```

Comment:
prelude exports `error!` now. Don't use `tracing::`.

Reply:
The prelude exports `error!` from our crate that accepts `logger` as the first argument. I believe I should use tracing instead - and I found multiple usages in other crates.

Reply:
Sorry, I forgot that my PR intorduces it and I havent merged it yet :P